Update README [skip-ci]

nbhuiyan opened the pull request:
This patch updates the README file to contain more up-to-date
information on building b9. There were some missing steps
that have now been addressed.

The README file contains instructions on how to build b9
on Ubuntu 16.04. The steps here make users update their
nodejs to the latest version to ensure all features being
used in the project are available.

Also, I switched the instructions to build out of source tree
as it is a cleaner approach in my opinion.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Update README [skip-ci]

This patch updates the README file to contain more up-to-date
information on building b9. There were some missing steps
that have now been addressed.

The README file contains instructions on how to build b9
on Ubuntu 16.04. The steps here make users update their
nodejs to the latest version to ensure all features being
used in the project are available.

Also, I switched the instructions to build out of source tree
as it is a cleaner approach in my opinion.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

rwy7 commented on README.md:
````diff
@@ -3,50 +3,76 @@
+
+### 3. Upgrade and create link for nodejs and install esprima
+
+```plaintext
````
Use `sh` codeblocks, not `plaintext`.

rwy7 commented on README.md:
````diff
@@ -3,50 +3,76 @@
+```plaintext
+npm install -g n \
+&& n latest \
+&& ln -s /usr/local/bin/node /usr/bin/node \
````
I am not crazy about using `n`. There are many ways to obtain node binaries. I think documenting our minimum version of nodejs in the main readme is a better way to go.

If Ubuntu's default packaged nodejs binary really is too old, we can suggest `n` as one way to manage nodejs versions. But even then, we should elaborate a little more on `n`, and provide a link.

Did we ever figure out what our minimum nodejs version is?

nbhuiyan replied:
The minimum version needed is version 4.5 since `Buffer.alloc()` was added in that version. I will document that in the main Readme.

rwy7 replied:
OK that would be awesome, thanks!

rwy7 replied:
OK so, (as you've said) on ubuntu 16.04 LTS, the default version really is too old.

Maybe we should link people to [this](https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions) document? This looks slightly more legit than `n`.

rwy7 commented on README.md:
````diff
@@ -3,50 +3,76 @@
+&& cd b9 \
+&& git submodule update --init \
+&& cd ..
+```
````
These last three steps are a no-op. Cloning the repository only requires a `git clone`. Can you mark this code block as `sh`?

rwy7 commented on README.md:
````diff
@@ -3,50 +3,76 @@
+* `ninja`
 
+To install the applications above, run the following command
+```plaintext
````
use `sh` not `plaintext`.

rwy7 reviewed (changes requested):
These instructions are sorely needed, but I think they're in the wrong place.  The toplevel README is purposefully general. We already have a document for [Ubuntu specific instructions](https://github.com/b9org/b9/blob/master/docs/SetupUbuntu.md). I would rather see you add this information to the ubuntu document, and link to it directly from the main README.

nbhuiyan committed:
Update build instructions [skip-ci]

This patch generalizes the main readme file and updates
the Ubuntu and OSX instruction page with more details.
I replicated these instructions and built successfully
on Mac OSX 10.13.3 and Ubuntu 16.04.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
@rwy0717 I addressed your feedback and made the necessary changes. It is ready for review.

rwy7 commented on docs/SetupOSX.md:
````diff
@@ -5,17 +5,60 @@ title: OSX Setup
+### 2. Clone the repository and get the submodules
+
+```sh
+cd ~ \
````
I'm not sure we need to tell people to `cd ~`, that's hopefully a one-time weirdness.

nbhuiyan replied:
Yeah, it is a mysterious one-time weirdness. I tried building it under Desktop today, and strangely it built just fine

rwy7 commented on docs/SetupUbuntu.md:
````diff
@@ -5,14 +5,78 @@ title: Ubuntu Setup
+
+```sh
+ln -s /usr/local/bin/nodejs /usr/bin/node
+```
````
People shouldn't need to create this link, or set this environment variable. Ubuntu users should be using the name `nodejs`. The build system should find either `nodejs` or `node`, as long as it's on the path.

I would rather see this section as:

---

### Upgrading node.js

For people running Ubuntu 16.04 and earlier, you may need to upgrade your nodejs installation. You can install newer node.js binaries from [NodeSource](https://nodesource.com/)'s PPA using:

```sh
curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash - \
&& sudo apt-get install -y nodejs
```

Check out [this](https://nodejs.org/en/download/package-manager) guide for more information on obtaining node.js packages.

rwy7 commented on docs/SetupUbuntu.md:
````diff
@@ -5,14 +5,78 @@ title: Ubuntu Setup
+You may verify that /usr/bin/node is indeed a link to version 4.5+ by running:
+```sh
+/usr/bin/node -v
+```
````
People should be testing that their `PATH` is correct, and running `nodejs -v`. This is the node executable cmake will find first.

nbhuiyan committed:
Update nodejs upgrade instructions [skip-ci]

Also-by: Robert Young <rwy0717@gmail.com>

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
@rwy0717 I have made the requested changes.

rwy7 commented:
Nice work!
